### PR TITLE
pss: Refactor. Step 3. Refactor outbox

### PR DESCRIPTION
**Pull request by aperonam**
~~**Note: This PR cannot be reviewed until PR #1734 is merged for the diff to make sense.**~~

This PR aims to refactor the outbox of Pss. Everything related to the outbox has been moved to its own package (outbox). A Start and Stop function has been added
Outbox tests have been splitted into whitebox and blackbox tests.

### CODE REVIEW

File: pss/outbox/outbox_test.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	forwardFail = true
+
+	err = testOutbox.Enqueue(testOutboxMessage)
+	if err != nil {
+		t.Fatalf("unexpected error enqueueing, %v", err)
+	}
+	//We wait for the forward function to fail
+	<-failedC
+
+	failedMessages := len(failed)
```
Comment (nolash) [inline pss/outbox/outbox_test.go]: Please consider whether this check is necessary.

File: pss/outbox/outbox_whitebox_test.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
+package outbox
+
+import (
+	"testing"
+
+	"github.com/ethersphere/swarm/pss/message"
+)
+
+func TestFullOutbox(t *testing.T) {
```
Comment (nolash) [inline pss/outbox/outbox_whitebox_test.go:13]: Please add comments describing what the tests actually do.

File: pss/outbox/outbox_whitebox_test.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	err = testOutbox.Enqueue(testOutboxMessage)
+	if err != nil {
+		t.Fatalf("unexpected error enqueueing, %v", err)
+	}
+	//As we haven't signaled processC, the messages are still in the outbox
+	err = testOutbox.Enqueue(testOutboxMessage)
+	if err != ErrOutboxFull {
+		t.Fatalf("unexpected error, got %v, wanted %v", err, ErrOutboxFull)
```
Comment (nolash) [inline pss/outbox/outbox_whitebox_test.go]: unexpected error _type_ ?

**Review by nolash — COMMENTED** [review]
I have only reviewed the changes mentioned in the description. I am assuming that all other code has been moved as-is - without any changes. If that's not the case then please let's review again and describe what additional changes exist.

File: pss/outbox/outbox_test.go (hunk 1 above)
Comment (aperonam) [reply pss/outbox/outbox_test.go]: I'm removing this

**Comment by aperonam** [comment]
> I have only reviewed the changes mentioned in the description. I am assuming that all other code has been moved as-is - without any changes. If that's not the case then please let's review again and describe what additional changes exist.

The only other change I haven't described is adding a mock.go file that allows  creating an outbox for tests without having to pass all the parameters (uses defaults).

File: pss/outbox/outbox_test.go [hunk 4]
```diff
@@ -3,15 +3,17 @@ package outbox_test
...
 	"testing"
+	"time"
 
 	"github.com/ethersphere/swarm/pss/message"
 	"github.com/ethersphere/swarm/pss/outbox"
 )
 
+const timeout = 2 * time.Second
+
 func TestOutbox(t *testing.T) {
```
Comment (nolash) [inline pss/outbox/outbox_test.go:15]: Also need a comment here ;)

File: pss/outbox/outbox.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+		}
+		return nil
+	default:
+		metrics.GetOrRegisterCounter("pss.enqueue.outbox.full", nil).Inc(1)
+		return ErrOutboxFull
+	}
+}
+
+//ProcessOutbox starts a routine that tries to forward messages present in the outbox queue
+func (o *Outbox) processOutbox() {
```
Comment (zelig) [inline pss/outbox/outbox.go:74]: look at this https://github.com/ethersphere/swarm/pull/1768/files#r325374003
the free function leak seems to be fixed but this very routine looks like will leak

File: pss/pss_test.go [hunk 6]
```diff
@@ -173,16 +172,13 @@ func TestAddressMatchProx(t *testing.T) {
...
-		for slot := range ps.outbox.process {
-			ps.outbox.free(slot)
-		}
-	}()
 	if err != nil {
 		t.Fatal(err.Error())
 	}
 
+	//Forwarding will fail. Since in this test we are not relying on forwarding, but just handling, the test is valid
+	ps.outbox.Start()
```
Comment (zelig) [inline pss/pss_test.go:179]: this doesnt look like its stopped

**Review by zelig — CHANGES_REQUESTED** [review]
LGTM but see comments

File: pss/outbox/outbox.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+	forwardFunc forwardFunction
+	queue       []*outboxMsg
+	slots       chan int
+	process     chan int
+	stopC       chan struct{}
+}
+
+type forwardFunction func(msg *message.Message) error
+
+var ErrOutboxFull = errors.New("outbox full")
```
Comment (nonsense) [inline pss/outbox/outbox.go:26]: One open question to me is whether this is a good API for PSS - what is the advantage of returning an `outbox full` error to clients, versus for example blocking the clients? To me it seems like this makes the API harder to use for clients at no apparent gains.

**Comment by janos** [comment]
@kortatu @nonsense I thought that this PR should be closed https://github.com/ethersphere/swarm/pull/1860#issuecomment-538336449 in favour to the newer one.

**Comment by kortatu** [comment]
Closed in favour of #1860 that includes this refactor.

**Pull request closed without merging**